File: C++-fast_double/benchmark.cpp
#include <time.h>
#include <iostream>
#include "fft.h"

#define SIZE (1<<LOG2FFTSIZE)

std::complex<double> xy[SIZE];
std::complex<double> xy_out[SIZE];

int main() {
    int i;
    double eltime;
    struct timespec gstart, gend;

    for(i=0; i<SIZE/2; i++) xy[i]= 1.;
    for(   ; i<SIZE  ; i++) xy[i]=-1.;

// FFT
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &gstart);
    fft *ff = new fft();
    for (i=0; i<FFT_REPEAT; i++) ff->calc(LOG2FFTSIZE, xy_out, xy);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &gend);
    eltime = 1000.*(gend.tv_sec - gstart.tv_sec) + (gend.tv_nsec - gstart.tv_nsec)/1000000.;
    printf("\n%6d piece(s) of %d pt FFT;  %9.5f ms/piece\n", FFT_REPEAT, 1<<LOG2FFTSIZE, eltime/FFT_REPEAT);

    for(i=0; i<6; i++) {
	printf("%3d %16.9f %16.9f %16.9f\n", i, xy_out[i].real(), xy_out[i].imag(), abs(xy_out[i]));
    }
    return 0;
}

File: Go/benchmark.go
package main

import (
    "fmt"
    "time"
    "fft"
)

const LOG2FFTSIZE uint= 12
const FFT_REPEAT int = 1000

const SIZE = (1<<LOG2FFTSIZE)
var xy [SIZE]complex128
var xy_out_fft [SIZE]complex128

func main() {
    var i int
    for i=0; i<SIZE/2; i++ { xy[i]= (1.+0i) }
    for    ; i<SIZE  ; i++ { xy[i]= (-1.+0i) }

// FFT
    var startTime = time.Now().UnixNano()
    for i=0; i<FFT_REPEAT; i++ { xy_out_fft = fft.Fft(LOG2FFTSIZE, xy) }
    var endTime = time.Now().UnixNano()

    fmt.Printf("%8d piece(s) of %d pt FFT;    %.3f ms/piece\n\n", FFT_REPEAT, SIZE, float64(endTime-startTime)/float64(1e6)/float64(FFT_REPEAT))

    for i=0; i<6; i++ {
	fmt.Printf("%d %f %f\n", i, real(xy_out_fft[i]), imag(xy_out_fft[i]))
    }
}
